lib/Runtime/Library/ScriptFunction.cpp

```diff
@@ -271,19 +271,8 @@ namespace Js
         Assert(this->functionInfo->GetFunctionBody() == newFunctionInfo);
         Assert(!newFunctionInfo->IsDeferred());
 
-        DynamicType * type = this->GetDynamicType();
-
-        // If the type is shared, it must be the shared one in the old function proxy
-
         this->functionInfo = newFunctionInfo->GetFunctionInfo();
 
-        if (type->GetIsShared())
```

Comment:
@curtisman If we're allowing shared types on function objects with non-default state, then we can't revert to the deferredPrototypeType here. I think we want to go ahead and change the entry point on the current shared type. Is there a reason we can't?

Reply:
My guess here is because originally, we would have swiitch out the functionInfo reference with the the undeferred function body in the scriptFunction.  So undeferring would only be fully done for one script function but not the other instance. (Thus can't change the shared entry point)

Since you have changed it that the switching is done in the function proxy, it's probably OK to do.  

But I wonder whether we need the line above any more as well.  Shouldn't this->functionInfo should equals to newFunctionInfo->GetFunctionInfo() always?

---
In reply to: [171666256](https://github.com/Microsoft/ChakraCore/pull/4748#discussion_r171666256) [](ancestors = 171666256)

Reply:
I think you're right. I've replaced the copy with an assert. I think the only other thing we need to do is to copy the deferredPrototypeType and undeferredFunctionType pointers on un-defer/re-defer now that we don't make new types for the new FunctionProxy.